cmd, core, eth, light, trie: add trie read caching layer

karalabe opened the pull request:
This is an alternative approach to https://github.com/ethereum/go-ethereum/pull/17873.

In short, this PR introduces a read cache into the `trie.Database` (previously it was a write cache/gc only). The reason behind this is twofold:

 * Storage tries are currently not cached. If subsequent blocks access the same read-only data, they keep hammering the disk to retrieve it. It's also a grieving vector since we can create transactions that try to max out storage trie reads (not dangerous at current gas levels, but sure annoying).
 * Executing pending transactions load parts of the account and storage tries into memory during their execution, but annoyingly dump all that information when a full block arrives. The next full block will however most probably share a lot of disk accesses with our pending state, so might as well reuse anything that we've already loaded.

My original attempt in https://github.com/ethereum/go-ethereum/pull/17873 cached entire storage tries, and introduced a fancy way to maximize useful data across transactions and blocks. Unfortunately, we have no means to measure the memory usage of that approach, neither easily control it. Alas, although it's cleaner and faster, it was not provably immune to DoS attacks.

This current approach in this PR adds the caching layer in between the database and our internal trie (same way we did for writes/pruning). This is suboptimal because we can only cache rlp encoded nodes, requiring reparsing them every time on access. On the flipside though, it avoid the disk reads the same way my old PR did, but also guarantees memory caps. As a bonus point, the data structure used avoids GC overhead on the millions of read-cached trie nodes, so it my actually be more performant with large enough caches.

---

### Benchmark results (purple = PR, blue = master):

Disk writes grow at the same rate as with master. The absolute value here is not relevant because I restarted the code on top of an existing chain, causing quite some blocks to be reprocessed, but different number on master and pr. The charts are after the system synced. Read wise we can see this pr saves about 75% of disk reads compared to master on mainnet (running with `--cache=2048`).

![screenshot from 2018-11-13 10-03-00](https://user-images.githubusercontent.com/129561/48399347-0ac41500-e72c-11e8-97da-1622a94330b8.png)

![screenshot from 2018-11-13 10-03-34](https://user-images.githubusercontent.com/129561/48399352-0c8dd880-e72c-11e8-982d-bc6eb816ae69.png)

Perhaps a more interesting metric is the propagated block processing time. This PR manages to cut new block times down by about 30%. You can see that the PR doesn't help old blocks that much during sync, since there's no preloading (pending transactions are the preloaders), however after sync is done, new import times are much lower than master ones. Our metrics library uses 1024 samples (blocks) for averaging, hence the couple hours it takes for the speedup to be fully visible on the charts.

![screenshot from 2018-11-13 10-10-58](https://user-images.githubusercontent.com/129561/48399495-7a3a0480-e72c-11e8-9ed4-4151b2f86468.png)

![screenshot from 2018-11-13 10-05-55](https://user-images.githubusercontent.com/129561/48399354-0e579c00-e72c-11e8-8110-d48ccc84c0b0.png)

holiman commented on core/state/database.go:
```diff
@@ -75,10 +75,10 @@ type Trie interface {
 // intermediate trie-node memory pool between the low level storage layer and the
 // high level trie abstraction.
-func NewDatabase(db ethdb.Database) Database {
+func NewDatabase(db ethdb.Database, cache int) Database {
```
The PR would have touched less files, and the interface might have been more clean, if you had created a separate `NewDatabaseWithCache(db ..., cache int)` and kept `NewDatabase(db)` as is, with cache set to `0`....

karalabe replied:
I did think about this. I just wanted to make sure I update all code paths. 0 is actually "infinite memory", not 0.

holiman commented on les/handler.go:
```diff
@@ -939,7 +939,7 @@ func (pm *ProtocolManager) handleMsg(p *peer) error {
 				var prefix string
 				if root, prefix = pm.getHelperTrie(req.Type, req.TrieIdx); root != (common.Hash{}) {
-					auxTrie, _ = trie.New(root, trie.NewDatabase(ethdb.NewTable(pm.chainDb, prefix)))
+					auxTrie, _ = trie.New(root, trie.NewDatabase(ethdb.NewTable(pm.chainDb, prefix), 1))
```
In most cases, you just put a `0` -- why `1` specifically here?

karalabe replied:
0 is infinite memory. In tests, I don't care. In live code however, we always want to have limits, so I set it to 1 MB here.

holiman commented on trie/database.go:
```diff
@@ -333,35 +347,49 @@ func (db *Database) insertPreimage(hash common.Hash, preimage []byte) {
 	// Content unavailable in memory, attempt to retrieve from disk
-	return db.diskdb.Get(hash[:])
+	enc, err := db.diskdb.Get(hash[:])
+	if err == nil || enc != nil {
```
Isn't it a bit fragile to use `||` here? Shouldn't you require `&&` ?

karalabe replied:
shit, that's flawed indeed

holiman commented on trie/database.go:
```diff
@@ -333,35 +347,49 @@ func (db *Database) insertPreimage(hash common.Hash, preimage []byte) {
+	// Retrieve the node from the dirty cache if available
 	db.lock.RLock()
-	node := db.nodes[hash]
+	dirty := db.nodes[hash]
```
`dirty` here simply means not-yet-written-to-disk, right?

holiman replied:
And if so, maybe we should rename `db.nodes` into `db.nodesNotOnDisk` ... or maybe not, just a thought

holiman commented:
Comparing `master` with this pr on `TestDifficulty/difficultyConstantinople.json/DifficultyTest1023`
```
        	difficulty_test.go:85: parent[time 28575090863 diff 541848385004890396 unclehash:1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347] child[time 28575090883 number 4300000] diff 541583810598149727 != expected 541583810598149727

        --- FAIL: TestDifficulty/difficultyRopsten.json/DifficultyTest1023 (0.00s)
        	difficulty_test.go:85: parent[time 9323269331 diff 7371607815720194386 unclehash:1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347] child[time 9323269351 number 4300000] diff 7368008397841424760 != expected 7368008397841426808
```
both parent time and child time seems off

holiman commented:
Do you get the same errors locally?

holiman commented:
Test 
```
[user@work go-ethereum]$ cat ./tests/testdata/BasicTests/difficultyRopsten.json  | grep DifficultyTest1023 -A8
 "DifficultyTest1023" : {
		"parentTimestamp" : "0x037f9b22f8",
		"parentDifficulty" : "0x482061c1ba2fb029",
		"parentUncles" : "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
		"currentTimestamp" : "0x037f9b230c",
		"currentBlockNumber" : "0x419ce0",
		"currentDifficulty" : "0x48175db581f86a33"
	},

```

karalabe commented:
Yeah, so I accidentally pushed some old version of the consensus test repo into this PR. Fixed it now.

karalabe commented:
@holiman I've rebased on master, renamed the `nodes` field to `dirties` to make it clearer what it is, and also added 4 metrics to be able to measure the hits/misses, data stores/loads from the read cache. These will be particularly handy when comparing a simple node vs. a light server to see how much strain is put on the caches.

fjl commented:
Changes look much cleaner now. Let's investigate if this change also allows disabling the generation-based caching.

karalabe commented:
Sure, but lets please do that in a followup PR. Removing the cache generations is a non negligible amount of change, after which at least a full sync is needed to validate it, which is +1 week. If this change is ok and works well, I don't think we should postpone it just to add more to it.

crackcomm commented:
@karalabe I would consider it unsafe in some circumstances but it's just being paranoid, is'nt it?

this is why I submitted #18118
